### ISSUE TITLE
Cant train with unsloth

### ISSUE THREAD

**Issue by Etherll** [edwko/OuteTTS#81]
Hello,
Thanks for the great models!
I'm trying to create an unsloth notebook for finetuning OuteTTS, but the output is very poor. It just produces weird noises and hallucinations.
I wonder if I've done something wrong. Here's the notebook: https://colab.research.google.com/drive/1S9pxP21TMezkAk55ih_6RajJQbpi_lqt#scrollTo=QmUBVEnvCDJv

**Comment by edwko** [comment]
Hey, nice work on the data preparation, up to that point, everything looks mostly correct. I noticed a few issues that could be causing problems:

In the inference section, you’re generating directly from `FastModel.get_peft_model`, but this model expects the windowed penalties ([see here](https://github.com/edwko/OuteTTS/blob/94e25d4f7fde64d416248c3394c9604321b75855/outetts/models/hf_model.py#L17)) to be applied. Instead, you should run inference directly through the `outetts` library using the `HF backend`.

The prompts being generated seem fine, but how you split the data for fine-tuning is important. Ideally, for SFT training, you should split the input randomly. For example, if the prompt has 10 words, you could split after 5 words: the first 5 words become the input, and the next 5 words become the completion. I didn't really see any kind of splitting happening in the colab notebook. Looking at your final loss, it's sitting around `2.6`, for SFT training, you usually want it to be below `1`. This suggests there might be issues with the data.

Let me know if this helps! Once it's working, feel free to open a PR with an example for fine-tuning. :)

**Comment by edwko** [comment]
Maybe also try different LoRA settings: something like r=32, lora_alpha=64, and a lower learning rate, around 4e-5. It would also be good to monitor the gradient norm if possible.

**Comment by Etherll** [comment]
Hey! I fixed the issue — thank you! It was an unsloth-related issue. I had to disable fast generation. I also added windowed penalties to the inference. I will test your idea of splitting later. Thanks for the suggestions! Here is the voice after fine-tuning. :D
https://voca.ro/168BrzgqlEzY
It's very good, but it talks a bit too fast. Can we adjust the generation settings to make it talk more slowly? Also, it ignores the emotive tags, such as <giggles>. It's in the dataset, and most TTS models were able to generate them. I think it's because Whisper ignores them, too.

**Comment by Etherll** [comment]
i opened a pr in unsloth/notebooks in case anyone is interested https://github.com/unslothai/notebooks/pull/35

**Comment by edwko** [comment]
Nice! Glad it works now. To make it talk slower, you should add a slower speaking voice as a reference speaker for the model. Since it's a voice cloning model, it will adjust to the speaker reference.  
For the voice effect tags, as mentioned, you should try splitting the input/target so you condition the model to properly finish it if it's present in the text input, for example:  

```
hello world <giggle>
...
```

then in the target:

```
...
<giggle> [features] [codes] [eow]
```

It should learn to align it separately. Try training with more data, and also make sure it's present in the prompts.

**Comment by omarkshoaib** [comment]
> i opened a pr in unsloth/notebooks in case anyone is interested [unslothai/notebooks#35](https://github.com/unslothai/notebooks/pull/35)

the notebook is not rendering could you upload it somewhere else ?

**Comment by Etherll** [comment]
> > i opened a pr in unsloth/notebooks in case anyone is interested [unslothai/notebooks#35](https://github.com/unslothai/notebooks/pull/35)
> 
> the notebook is not rendering could you upload it somewhere else ?

you can use download it or use colab https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Oute_TTS_(1B).ipynb
